docs/source/data/fetching-data.mdx

```diff
@@ -84,48 +84,48 @@ If you want to add batching, deduplication, or caching to your data source class
 
 DataLoader provides a memoization cache, which avoids loading the same object multiple times during a single GraphQL request ([much like one of `RESTDataSource`'s caching layers](./fetching-rest#get-requests-and-responses)). It also combines loads during a single event loop tick into a batched request that fetches multiple objects at once.
 
-DataLoader instances are pre-request, so if you use a DataLoader in your data source, ensure you **create a new instance of that class with every request**/ :
+DataLoader instances are per-request, so if you use a DataLoader in your data source, ensure you **create a new instance of that class with every request** :
+
+<MultiCodeBlock>
 
 ```ts
 import DataLoader from 'dataloader';
 
 class ProductsDataSource {
   private dbConnection;
 
-  constructor() {
-    this.dbConnection = this.initializeDBConnection();
-  }
-
-  async initializeDBConnection() {
-    // set up our database details, instantiate our connection,
-    // and return that database connection
-    return dbConnection;
-  }
-
   private batchProducts = new DataLoader(async (ids) => {
-    const progressList = await db.fetchAllKeys(ids);
-    return ids.map((id) => progressList.find((progress) => progress.id === id));
+    const productList = await db.fetchAllKeys(ids);
+    return ids.map((id) => productList.find((product) => product.id === id));
```

Comment:
Technically this introduces unnecessary quadratic behavior.  Imagine that ids is 1000 elements long. That means productList generally will be too.  Then each `find` call takes on average 500 comparisons to find the id it is looking for, meaning you end up with 500,000 comparisons!  A better approach would be to put the returned items in some sort of map structure so that you just need to do one lookup per ID at the end.

It's not the end of the world to have this in an example but certainly it's nice to obey best practices. Fortunately in this example, we have the superpower that we can assume our underlying APIs do whatever is most useful :) So we could do something like

```
const productsById = await db.fetchAllKeys(ids);
return ids.map((id) => productsById.get(id));
```